contracts/utils/structs/MerkleTree.sol

```diff
@@ -0,0 +1,187 @@
+
+        // Build the different hashes in a zero-filled complete tree
+        bytes32 currentZero = zero;
+        for (uint32 i = 0; i < depth; ++i) {
+            Arrays.unsafeAccess(self.zeros, i).value = currentZero;
+            currentZero = fnHash(currentZero, currentZero);
+        }
+
+        // Insert the first root
+        Arrays.unsafeAccess(self.roots, 0).value = currentZero;
+    }
+
+    /**
+     * @dev Insert a new leaf in the tree, compute the new root, and store that new root in the history.
+     */
+    function insert(TreeWithHistory storage self, bytes32 leaf) internal returns (uint256) {
+        // Cache read
+        uint256 depth = self.zeros.length;
+        function(bytes32, bytes32) view returns (bytes32) fnHash = self.fnHash;
+
+        // Get leaf index
+        uint256 leafIndex = self.nextLeafIndex++;
+
+        // Check if tree is full.
+        if (leafIndex >= 1 << depth) Panic.panic(Panic.RESOURCE_ERROR);
+
+        // Rebuild branch from leaf to root
+        uint256 currentIndex = leafIndex;
+        bytes32 currentLevelHash = leaf;
+        for (uint32 i = 0; i < depth; i++) {
+            // Reaching the parent node, is currentLevelHash the left child?
+            bool isLeft = currentIndex % 2 == 0;
+
+            // If so, next time we will come from the right, so we need to save it
+            if (isLeft) {
+                Arrays.unsafeAccess(self.sides, i).value = currentLevelHash;
+            }
+
+            // Compute the node hash by hashing the current hash with either:
+            // - the last value for this level
+            // - the zero for this level
+            currentLevelHash = fnHash(
+                isLeft ? currentLevelHash : Arrays.unsafeAccess(self.sides, i).value,
+                isLeft ? Arrays.unsafeAccess(self.zeros, i).value : currentLevelHash
+            );
+
+            // Update node index
+            currentIndex >>= 1;
+        }
+
+        // Record new root
+        self.currentRootIndex = (self.currentRootIndex + 1) % self.roots.length;
+        Arrays.unsafeAccess(self.roots, self.currentRootIndex).value = currentLevelHash;
+
+        return leafIndex;
+    }
+
+    /**
+     * @dev Tree's depth (set at initialization)
+     */
+    function getDepth(TreeWithHistory storage self) internal view returns (uint256) {
+        return self.zeros.length;
+    }
+
+    /**
+     * @dev History length (set at initialization)
+     */
+    function getLength(TreeWithHistory storage self) internal view returns (uint256) {
+        return self.roots.length;
+    }
+
+    /**
+     * @dev Return the current root of the tree.
+     */
+    function getLastRoot(TreeWithHistory storage self) internal view returns (bytes32) {
+        return Arrays.unsafeAccess(self.roots, self.currentRootIndex).value;
+    }
+
+    /**
+     * @dev Look in root history,
+     */
+    function isKnownRoot(TreeWithHistory storage self, bytes32 root) internal view returns (bool) {
```

Comment:
What are we trying to achieve with this function?

I think the length of the tree reduces the utility of this function since it will start reporting `false` after the length has been reached and the values are started to get overridden. If the function is on-chain, I think it's still useful if the tree keeps values for a certain period, but doesn't sound like that's going the be everyone's use case.

In my opinion, a function like this should be callable in `O(1)` and keep knowing all the roots that have been registered. So it'll be better if we have a getter to `self.roots[i]` and we let developers implement their `isKnownRoot` (perhaps we do it through an extension and just use a Bitmap for it).

However, if we add a getter for `self.roots[i]` we should be careful of allowing arbitrary length, as we may be adding a getter for the whole. I'd suggest restricting root's length to `uint8`